### ISSUE TITLE
Fedora rpm spec

### ISSUE THREAD

**Issue by tieugene** [joshgoebel/keyszer#64]
Here is the spec-file for creating rpm package for [Fedora 36](https://koji.fedoraproject.org/koji/taskinfo?taskID=89385741).
CentoOS/RHEL 8-9 require a little bit more job, but it is possible too (on demand).
I not tested it on myself because cannot decide what to do after installation - setup system-wide keyszer.service (BTW it is 404 from README.md) with root or with separate user (and how to use it after start?), or create user systemd unit, or... README.md is not too clear.
This spec can be placed right into root of sources. Then any user can `rpmbuild -ta keyszer-0.6.0.tar.gz` and get the rpm.

```spec
# keyszer.spec
Name:		keyszer
Version:	0.6.0
Release:	1%{?dist}
License:	GPLv3
Summary:	X keyboard remappring tool
URL:		https://github.com/joshgoebel/%{name}
Source0:	https://github.com/joshgoebel/%{name}/archive/%{name}-%{version}.tar.gz
BuildRequires:	pyproject-rpm-macros
BuildArch:	noarch

%description
Keyszer is a smart key remapper for Linux (and X11) written in Python.
It's similar to xmodmap but allows far more flexible remappings.


%prep
%autosetup


%generate_buildrequires
%pyproject_buildrequires -w


%build
%pyproject_wheel


%install
%pyproject_install
%pyproject_save_files keyszer


%files -f %{pyproject_files}
%doc README.md example/*
%{_bindir}/%{name}


%changelog
* Mon Jul 11 2022 TI_Eugene <ti.eugene@gmail.com> - 0.6.0-1
- Initial build
```

**Comment by joshgoebel** [comment]
We want it to be a system (but non-root) service.. a dedicated user is best.  This is the service file I have, but I haven't tested it yet.

```
[Unit]
Description=keyszer

[Service]
Type=simple
KillMode=process
ExecStart=keyszer --quiet --watch
Restart=on-failure
RestartSec=3
Environment=DISPLAY=:0
User=keymapper
Group=input

[Install]
WantedBy=graphical.target
```

**Comment by tieugene** [comment]
> a dedicated user is best.

Next question - what is group of this user? As for Fedora `useradd keymapper` creates as user as group with this name

**Comment by joshgoebel** [comment]
The group really doesn't matter since we're doing per-user permissions, but a paired group/used like keymapper/keymapper is fine.

**Comment by tieugene** [comment]
> a paired group/used like keymapper/keymapper is fine.

What to create yet another group for?
I think built-in `nobody` group is the best for this.

**Comment by joshgoebel** [comment]
That's probably fine to. If you want to do some actual testing running as a service that'd be great.

**Comment by tieugene** [comment]
> That's probably fine to. If you want to do some actual testing running as a service that'd be great.

I'm trying right now on myself (Fedora 36 @ MacBook Pro mid 2012):
https://github.com/tieugene/keyszer/tree/rh/contrib
This is fork with special branch for RH-based (`rh`).
Currently the service work but do nothing.
Log tail:
```log
июл 12 17:20:30 localhost.localdomain keyszer[3443]: (DD) on_key BACKSPACE press
июл 12 17:20:30 localhost.localdomain keyszer[3443]: (OO) press BACKSPACE
июл 12 17:20:30 localhost.localdomain keyszer[3443]: (II) in BACKSPACE (release)
июл 12 17:20:30 localhost.localdomain keyszer[3443]: (EE) Can't connect to display ":0": b'No protocol specified\n'
```

**Comment by tieugene** [comment]
So, seems system-wide service will not work with user's display.
I decide to user user service.
This works ok: https://github.com/tieugene/keyszer/tree/rh/contrib (without `not_works` subdir).
I can make PR if this will be usual

**Comment by joshgoebel** [comment]
> So, seems system-wide service will not work with user's display.

I thought this was explained in the docs... but seems not... you need to give the keymapper user permission to access X:

```
xhost +SI:localuser:keymapper
```

That has to be run as the user who is using X. 

> I decide to use user service.

Please so the README on why this is a dangerous decision.

**Comment by joshgoebel** [comment]
> I can make PR if this will be usual

Not interested in user service unless you have an answer to the security issues...

**Comment by tieugene** [comment]
> ```
> xhost +SI:localuser:keymapper
> ```

Now it works with system-wide systemd unit, thank you.
Almost full-auto, but:
1. `/dev/uinput` is not writable by `keymapper` dispite the [rules](https://github.com/tieugene/keyszer/blob/rh/contrib/system/99-keymapper-acl.rules) (but `/dev/input/event*` OK). Repeating this command with sudo set required ACL.
2. The service is system-wide but cannot be started before user logged in (no X - no job)

**Comment by joshgoebel** [comment]
> /dev/uinput is not writable by keymapper

Did you rerun the rules/retrigger? I forget how...

> The service is system-wide but cannot be started before user logged in (no X - no job)

Why not?  It should just pass keystrokes thru without harm if there is no X around...

**Comment by tieugene** [comment]
> > /dev/uinput is not writable by keymapper
> 
> Did you rerun the rules/retrigger? I forget how...

I'll try but this is not right way, IMHO.

> > The service is system-wide but cannot be started before user logged in (no X - no job)
> 
> Why not? It should just pass keystrokes thru without harm if there is no X around...

Because `cannot connect to display`. I tried to run service, logout and login again and have bit log with these errors.

**Comment by joshgoebel** [comment]
> auto xhost +SI...

I'm not sure that's in scope... though it should be in our docs... might be in scope for Kinto.sh though.

**Comment by tieugene** [comment]
> Why would you even want to run the service in verbose mode at all? I'm not sure why we need a toggle?

 You are right - verbose is not usual for production. Feature request declined.

**Comment by tieugene** [comment]
> > auto xhost +SI...
> 
> I'm not sure that's in scope... though it should be in our docs... might be in scope for Kinto.sh though.

If you mean [USING_WITH_KINTO.md](https://github.com/joshgoebel/keyszer/blob/main/USING_WITH_KINTO.md) it is not correct.
I think the problem is:
1. if we starts `keyszer.service` as system[-wide] service then this has no sense _after_ X started but _before_ user logged in (login screen): `xhost +si:localuser:keymapper`. May be I'm wrong because not familiar how X works.
2. if we starts the service as user (`systemctl --user ...`) then this not works: `User=keymapper, Group=input`

**Comment by joshgoebel** [comment]
No I meant in general.  Kinto already handles the X permissions because it's running as root, no?  I assume someone dropping us into a Kinto install (for now) would continue running as root.  Long-term this is something Kinto should change when they switch to keyszer.

> then this has no sense after X started but before user logged in

Huh?  Can you not give permission to the X display itself BEFORE a user has logged in?  If this already works in Kinto I would assume it works because you CAN do exactly that - and that because the target is multi-user that we're being started after X is running... so there should be something to give permission to.

CC @rbreaves

**Comment by tieugene** [comment]
According to keyszer's log it cannot get display right after user log in. Because this resets user's X permissions.
So the last problem is how to `xhost +si...` by default during creating `~/.Xauthority`

**Comment by tieugene** [comment]
> I wonder how Kinto solves this problem then... esp consider that xkeysnail is VERY VERY unhappy if it can't access the X display and will simply crash... where-as keyszer just shows an error and will keep waiting to see if the X display becomes available later.

Anyway seems we win:
- `keymapper` user: &check;
- `/dev/*` permissions: &check;
- `xhost +si...` (auto): &check;
- `keyszer.service` (system): &check; (enabled, starts on system boot and works _quietly_).
It's time to `rpmbuild` this

**Comment by tieugene** [comment]
Something bad with `--watch` (sources downloaded as Release tarball):
```
июл 13 21:44:44 localhost.localdomain systemd[1]: Started keyszer.service - keyszer.
июл 13 21:44:45 localhost.localdomain keyszer[2209]: keyszer v0.6.0
июл 13 21:44:45 localhost.localdomain keyszer[2209]: (--) WATCH: Watching for new devices to hot-plug.
июл 13 21:44:45 localhost.localdomain keyszer[2209]: (--) Autodetecting keyboards (--device not specified)
июл 13 21:44:45 localhost.localdomain keyszer[2209]: Traceback (most recent call last):
июл 13 21:44:45 localhost.localdomain keyszer[2209]:   File "/usr/bin/keyszer", line 8, in <module>
июл 13 21:44:45 localhost.localdomain keyszer[2209]:     sys.exit(main())
июл 13 21:44:45 localhost.localdomain keyszer[2209]:   File "/usr/lib/python3.10/site-packages/keyszer/cli.py", line 123, in main
июл 13 21:44:45 localhost.localdomain keyszer[2209]:     main_loop(args.devices, args.watch)
июл 13 21:44:45 localhost.localdomain keyszer[2209]:   File "/usr/lib/python3.10/site-packages/keyszer/input.py", line 170, in main_loop
июл 13 21:44:45 localhost.localdomain keyszer[2209]:     devices, device_filter, inotify)
июл 13 21:44:45 localhost.localdomain keyszer[2209]: NameError: name 'device_filter' is not defined. Did you mean: 'DeviceFilter'?
июл 13 21:44:45 localhost.localdomain systemd[1]: keyszer.service: Main process exited, code=exited, status=1/FAILURE
июл 13 21:44:45 localhost.localdomain systemd[1]: keyszer.service: Failed with result 'exit-code'.
```

**Comment by tieugene** [comment]
So, [it works](https://github.com/tieugene/keyszer/tree/rh/contrib/system) (for me) - system-wide, full-auto, as ordered.
Ready rpm can be [downloaded](https://koji.fedoraproject.org/koji/taskinfo?taskID=89468561)

**Comment by joshgoebel** [comment]
> Something bad with --watch

0.6 is VERY old, things are still moving very fast. If you try `main` branch I think that issue is  fixed there.  I'm not seeing it here.

**Comment by joshgoebel** [comment]
@tieugene I don't really know much about RPM distros... do people on those distros expect to build their own RPMs for stuff like this - should we try and get in an official repository somewhere eventually?  Or should we have a `rpmbuild` folder as part of this project or something?

I plan to submit to AUR (Arch) once I have things a bit more stable, and I understand how that system works...

**Comment by tieugene** [comment]
Everybody who use Fedora can install `rpmbuild` from official repo and make rpm.
Everybody who use rpm-based distro can tune .spec and make rpm for his distro.
Publishing ready rpms on the site is not good idea because it must be constantly synced with official repo (dependencies).
Better way is to try to submit application into official repo. As Fedora maintainer I can do this when it will be less or more maintainable (now documentation is not good).

I propose to make `/contrib` folder and collect there all things to make packages. E.g. `/contrib/redhat` with spec and distro-specific add-ons.

**Comment by joshgoebel** [comment]
> I propose to make /contrib folder and collect there all things to make packages. E.g. /contrib/redhat 

That sounds like a reasonable start.

> now documentation is not good).

Yeah, I'd love help with that... I've been more focused on the keymapper itself so far and correctness. I definitely do think docs are super important.

**Comment by tieugene** [comment]
> > I propose to make /contrib folder and collect there all things to make packages. E.g. /contrib/redhat
> 
> That sounds like a reasonable start.
> 
> > now documentation is not good).
> 
> Yeah, I'd love help with that... I've been more focused on the keymapper itself so far and correctness. I definitely do think docs are super important.

I think good documentation is not important but mandatory. As for me I never try to submit new package if documentation is not clear.
BTW I propose to split main README into parts:
- changelog - to Changelog.[txt/md]
- usage - as main README or separate md
- installation - into Install.md or something like this
And idea is that when (?/!) keyszer will be in official repos there is not sense to include installation and changelog into package documentation.
Maybe creating special `doc/` folder for all 2nd-level texts will be usual

**Comment by joshgoebel** [comment]
> changelog - to Changelog.[txt/md]

That's fine and to be expected - long term. I see where we are right now (trying to get people to switch away from the other unmaintained project) as unique - and why I list the large amount of progress I've been making so far in README - it's important to know.

> installation - into Install.md or something like this

If the instructions are going to continue to be as long as they are now than you're probably right - so long as the README still links to `INSTALL.md`...

**Comment by tieugene** [comment]
> > Something bad with --watch
> 
> 0.6 is VERY old, things are still moving very fast. If you try `main` branch I think that issue is fixed there. I'm not seeing it here.

Right now from freshest code (commit #4ddf5ff):
```
июл 16 13:13:43 macbook systemd[1]: Started keyszer.service - keyszer.
июл 16 13:13:49 macbook keyszer[18511]: Traceback (most recent call last):
июл 16 13:13:49 macbook keyszer[18511]:   File "/usr/lib/python3.10/site-packages/keyszer/input.py", line 162, in device_change
июл 16 13:13:49 macbook keyszer[18511]:     if registry.cares_about(device):
июл 16 13:13:49 macbook keyszer[18511]:   File "/usr/lib/python3.10/site-packages/keyszer/devices.py", line 68, in cares_about
июл 16 13:13:49 macbook keyszer[18511]:     return self._filter(device)
июл 16 13:13:49 macbook keyszer[18511]: TypeError: 'DeviceFilter' object is not callable
```

**Comment by joshgoebel** [comment]
Should be fixed  41085d147a3728bfbec622276e45414234f3bc43
